man8/cache_restore.8
.TH CACHE_RESTORE 8 "Thin Provisioning Tools" "Red Hat, Inc." \" -*- nroff -*-
.SH NAME
cache_restore \- restore cache metadata file to device or file

.SH SYNOPSIS
.B cache_restore
.RB [ options ]
.RB -i
.I {device|file}
.RB -o
.I {device|file}

.SH DESCRIPTION
.B cache_restore
restores cache metadata created by the respective
device-mapper target dumped into an XML formatted (see
.BR cache_dump(8) )
.I file
, which optionally can be preprocessed before the restore to another
.I device
or
.I file.
If restored to a metadata
.I device
, the metadata can be processed by the device-mapper target.

This tool cannot be run on live metadata.

.IP "\fB\-i, \-\-input\fP \fI{device|file}\fP"
Input file or device with metadata.

.IP "\fB\-o, \-\-output\fP \fI{device|file}\fP"
Output file or device for repaired binary metadata.  If a file is used
then it must be preallocated, and large enough to hold the metadata.

.IP "\fB{\-\-debug-override-metadata-version}\fP \fI<integer>\fP"
ONLY FOR DEBUGGING PURPOSES:
.br
override the version stored in the metadata.

.IP "\fB\-h, \-\-help\fP"
Print help and exit.

.IP "\fB\-V, \-\-version\fP"
Output version information and exit.

.SH EXAMPLE
Restores the XML formatted cache metadata on file
.B metadata
to logical volume /dev/vg/metadata for further processing by the
respective device-mapper target:
.sp
.B cache_restore -i metadata -o /dev/vg/metadata

.SH DIAGNOSTICS
.B cache_restore
returns an exit code of 0 for success or 1 for error.

.SH SEE ALSO
.B cache_dump(8)
.B cache_check(8)
.B cache_repair(8)

.SH AUTHOR
Joe Thornber <ejt@redhat.com>
.br
Heinz Mauelshagen <HeinzM@RedHat.com>


CHANGES
v0.6.2
======

- Fix bug in thin_delta
- Fix recent regression in thin_repair.
- Force g++-98 dialect
- Fix bug in thin_trim

v0.6.1
======

Fix bug in thin_dump when using metadata snaps.

v0.6.0
======

thin_ls

v0.5.6
======

era_invalidate may be run on live metadata if the --metadata-snap option is given.

v0.5.5
======

You may now give the --metadata_snap option to thin_delta without specifying where the snap is.

Update man pages to make it clearer that most tools shouldn't be run on live metadata.

Fix some bugs in the metadata reference counting for thin_check.

v0.5.1
======

Fix bug where the tools would crash if given a very large metadata
device to restore to.

v0.5
====

- thin_delta, thin_trim
- --clear-needs-check flag for cache_check
- space map checking for thin check

v0.4
====

- All tools switch to using libaio.  This gives a large performance
  boost, especially to the write focused tools like thin_restore.

- Added a progress monitor to thin_restore, cache_restore and era_restore

- Added a --quiet/-q option to *_restore to turn off the progress bar

- Removed variable hint size support from cache tools.  The kernel
  still only supports a fixed 32bit width.  This will have a side
  effect of reducing the executable sizes due to less template
  instatiation.

- Tools rolled into a single executable to save space.

- Fixed some bugs when walking bitsets (possibly effecting cache_dump
  and cache_check).

caching/cache_restore.cc
#include "version.h"

#include "base/output_file_requirements.h"
#include "caching/commands.h"
#include "caching/metadata.h"
#include "caching/restore_emitter.h"
#include "caching/xml_format.h"
#include "persistent-data/file_utils.h"

#include <boost/lexical_cast.hpp>
#include <fstream>
#include <getopt.h>
#include <iostream>
#include <libgen.h>
#include <string>

using namespace boost;
using namespace caching;
using namespace persistent_data;
using namespace std;

//----------------------------------------------------------------

namespace {
	size_t get_file_length(string const &file) {
		struct stat info;
		int r;

		r = ::stat(file.c_str(), &info);
		if (r)
			throw runtime_error("Couldn't stat backup path");

		return info.st_size;
	}

	auto_ptr<progress_monitor> create_monitor(bool quiet) {
		if (!quiet && isatty(fileno(stdout)))
			return create_progress_bar("Restoring");
		else
			return create_quiet_progress_monitor();
	}

	struct flags {
		flags()
			: metadata_version(1),
			  override_metadata_version(false),
			  clean_shutdown(true),
			  quiet(false) {
		}

		optional<string> input;
		optional<string> output;

		uint32_t metadata_version;
		bool override_metadata_version;
		bool clean_shutdown;
		bool quiet;
	};

	int restore(flags const &fs) {
		try {
			block_manager<>::ptr bm = open_bm(*fs.output, block_manager<>::READ_WRITE);
			metadata::ptr md(new metadata(bm, metadata::CREATE));
			emitter::ptr restorer = create_restore_emitter(md, fs.clean_shutdown);

			if (fs.override_metadata_version) {
				cerr << "overriding" << endl;
				md->sb_.version = fs.metadata_version;
			}

			check_file_exists(*fs.input);
			ifstream in(fs.input->c_str(), ifstream::in);

			auto_ptr<progress_monitor> monitor = create_monitor(fs.quiet);
			parse_xml(in, restorer, get_file_length(*fs.input), *monitor);

		} catch (std::exception &e) {
			cerr << e.what() << endl;
			return 1;
		}

		return 0;
	}
}

//----------------------------------------------------------------

cache_restore_cmd::cache_restore_cmd()
	: command("cache_restore")
{
}

void
cache_restore_cmd::usage(std::ostream &out) const
{
	out << "Usage: " << get_name() << " [options]" << endl
	    << "Options:" << endl
	    << "  {-h|--help}" << endl
	    << "  {-i|--input} <input xml file>" << endl
	    << "  {-o|--output} <output device or file>" << endl
	    << "  {-q|--quiet}" << endl
	    << "  {-V|--version}" << endl
	    << endl
	    << "  {--debug-override-metadata-version} <integer>" << endl
	    << "  {--omit-clean-shutdown}" << endl;
}

int
cache_restore_cmd::run(int argc, char **argv)
{
	int c;
	flags fs;
	char const *short_opts = "hi:o:qV";
	option const long_opts[] = {
		{ "debug-override-metadata-version", required_argument, NULL, 0 },
		{ "omit-clean-shutdown", no_argument, NULL, 1 },
		{ "help", no_argument, NULL, 'h'},
		{ "input", required_argument, NULL, 'i' },
		{ "output", required_argument, NULL, 'o'},
		{ "quiet", no_argument, NULL, 'q'},
		{ "version", no_argument, NULL, 'V'},
		{ NULL, no_argument, NULL, 0 }
	};

	while ((c = getopt_long(argc, argv, short_opts, long_opts, NULL)) != -1) {
		switch(c) {
		case 0:
			fs.metadata_version = lexical_cast<uint32_t>(optarg);
			fs.override_metadata_version = true;
			break;

		case 1:
			fs.clean_shutdown = false;
			break;

		case 'h':
			usage(cout);
			return 0;

		case 'i':
			fs.input = optional<string>(string(optarg));
			break;

		case 'o':
			fs.output = optional<string>(string(optarg));
			break;

		case 'q':
			fs.quiet = true;
			break;

		case 'V':
			cout << THIN_PROVISIONING_TOOLS_VERSION << endl;
			return 0;

		default:
			usage(cerr);
			return 1;
		}
	}

	if (argc != optind) {
		usage(cerr);
		return 1;
	}

        if (!fs.input) {
		cerr << "No input file provided." << endl << endl;
		usage(cerr);
		return 1;
	}

	if (fs.output)
		check_output_file_requirements(*fs.output);

	else {
		cerr << "No output file provided." << endl << endl;
		usage(cerr);
		return 1;
	}

	return restore(fs);
}

//----------------------------------------------------------------
